generate_workspace/src/main/java/com/google/devtools/build/workspace/maven/VersionResolver.java

```diff
@@ -0,0 +1,109 @@
+package com.google.devtools.build.workspace.maven;
+
+
+import com.google.common.collect.Lists;
+import org.eclipse.aether.RepositorySystem;
+import org.eclipse.aether.RepositorySystemSession;
+import org.eclipse.aether.artifact.Artifact;
+import org.eclipse.aether.repository.RemoteRepository;
+import org.eclipse.aether.resolution.VersionRangeRequest;
+import org.eclipse.aether.resolution.VersionRangeResolutionException;
+import org.eclipse.aether.resolution.VersionRangeResult;
+
+import java.util.List;
+
+import static com.google.devtools.build.workspace.maven.ArtifactBuilder.InvalidArtifactCoordinateException;
+
+/**
+ * Given a Maven coordinate with a version specification resolves the version of the coordinate in a
+ * similar fashion as Maven. Version specifications can include hard and soft pins as well as various
+ * forms of version ranges. When given a version range, Maven selects the highest available version.
+ * For both soft and hard pins, e.g. "4.2" or "[4.2]", Maven defaults to the pinned version.
+ */
+class VersionResolver {
+
+  private final List<RemoteRepository> remoteRepositories;
+  private final RepositorySystem repositorySystem;
+  private final RepositorySystemSession repositorySystemSession;
+
+  private VersionResolver(RepositorySystem repositorySystem,
+                          RepositorySystemSession systemSession,
+                          List<RemoteRepository> remoteRepositories) {
+    this.repositorySystem = repositorySystem;
+    this.remoteRepositories = remoteRepositories;
+    this.repositorySystemSession = systemSession;
+  }
+
+  /**
+   * Given a maven coordinate and its version specifications, selects the highest version if it is a version range
+   * or returns the pinned version if is a hard or soft pin.
+   */
+  String resolveVersion(String groupId, String artifactId, String versionSpec)
+      throws InvalidArtifactCoordinateException {
+
+    Artifact artifact = ArtifactBuilder.fromCoords(groupId, artifactId, versionSpec);
+    VersionRangeRequest rangeRequest = new VersionRangeRequest(artifact, remoteRepositories, null);
+
+    VersionRangeResult rangeResult;
+    try {
+      rangeResult = repositorySystem.resolveVersionRange(repositorySystemSession, rangeRequest);
```

Comment:
> If I say 3.0, no one conflicts with 3.0, 3.0 exists... 

How do we determine that 3.0 exists without checking the Maven Central? 

Suppose there is some artifact `a:b:3.12` we want. Now, suppose we create a mirror of maven central. This `a:b:3.12` exists within Maven Central. However, there is no `a:b:3.12` in the mirror, but there is a `a:b:3.13`. Using the the above heuristic in lieu of a network request, we would select `a:b:3.12.` Later, when we try to fetch the artifact, we would find that there is no `a:b:3.12`, and then skip resolving its transitive dependencies. 

This is incorrect behavior. "3.10" is a soft pin, rather than a hard pin. In the situation, there is no 3.10, it should still be able to use 3.11, 3.12, 3.13, etc.

> How about using some of the existing "find a version" code and, if that can't find anything, we do a network request?

This is very much a tradeoff between correctness and performance. IMO, until we can match maven/aether one to one, I think we should err on the side of over correctness. We can freely tune things afterwards.